### ISSUE TITLE
Deoplete wrongfully complains about outdated neovim python package.

### ISSUE THREAD

**Issue by purpleP** [Shougo/deoplete.nvim#347]
[deoplete] Current neovim-python module version: ['0.1.10']
[deoplete] deoplete.nvim requires neovim-python 0.1.8+.
[deoplete] Please update neovim-python by "pip3 install --upgrade neovim"

AFAIK 10 > 8, right? =)

**Comment by purpleP** [comment]
Okay, Seems like i fixed this.

diff --git a/autoload/deoplete/init.vim b/autoload/deoplete/init.vim
index a3cd3c8..c4067ff 100644
--- a/autoload/deoplete/init.vim
+++ b/autoload/deoplete/init.vim
@@ -61,8 +61,9 @@ function! deoplete#init#_channel() abort "{{{
   endtry
 
   " neovim module version check.
+
   if empty(g:deoplete#_neovim_python_version) ||
-        \ sort(g:deoplete#_neovim_python_version)[-1] < '0.1.8'
+	  \ deoplete#init#cmp_versions(sort(g:deoplete#_neovim_python_version)[-1], '0.1.8')
     call deoplete#util#print_error(
           \ 'Current neovim-python module version: ' .
           \  string(g:deoplete#_neovim_python_version))
@@ -225,3 +226,24 @@ function! deoplete#init#_context(event, sources) abort "{{{
 endfunction"}}}
 
 " vim: foldmethod=marker
+
+function! deoplete#init#cmp_versions(a, b)
+	let l:a_nums = split(a:a, '.')
+	let l:b_nums = split(a:b, '.')
+	" is python style unpacking possible?
+	for l:nums in deoplete#init#zip_longest('0', a_nums, b_nums)
+		if l:nums[0] < l:nums[1]
+			return 1
+		endif
+	endfor
+	return 0
+endfunction
+
+
+function! deoplete#init#zip_longest( defaultValue, ... )
+    let l:result = []
+    for l:i in range(max(map(copy(a:000), 'len(v:val)')))
+	call add(l:result, map(copy(a:000), 'get(v:val, l:i, a:defaultValue)'))
+    endfor
+    return l:result
+endfunction

**Comment by tweekmonster** [comment]
Have you updated recently?  This should be fixed in #342

**Comment by purpleP** [comment]
@tweekmonster Oh, okay, although this is clearly an overkill
```vim
for i in range(l)
     let d += (a[i] - b[i]) * pow(100, l - i - 1)
endfor
```
and this should be more concise and have better perfomance in theory. Using python notation
```python
def cmp(v1, v2):
    orderings = (
        a is None or a < b
        for a, b in zip_longest(None, *map(lambda v: v.split('.'), (v1, v2)))
        if a != b
    )
    ordering = next(orderings, None)
    return cmp(len(p1), len(p2) if ordering is None else 1 if ordering else -1
    
```
Something like that.

**Comment by tweekmonster** [comment]
@purpleP Clearly overkill??  The results are verifiable and uses a single loop that iterates only 3 times in a realistic scenario.  It's beautifully simple in my opinion.  The only thing I can see to optimize its performance is to not use `pow()` and use a list of multipliers instead.  Or even just not use a loop at all.  But, I don't think the microseconds it would shave off would win me any praise.

I can't get the python snippet you posted to work.  Just from looking, I'd say that the recursion and use of generators is beyond overkill for version comparison.  For simple boolean version comparison in Python, this is basically all there is to it:

```python
def cmp(v1, v2):
    return [int(x) for x in v1.split('.')] < [int(x) for x in v2.split('.')]
```

It'll only compare what can be compared in both lists.  `[5, 2] > [1, 2, 3, 4]` is `True`, `[1, 2] > [1, 2, 3, 4]` is `False`.  But, if you want to actually know how many versions apart the two are, you may need to do something closer to what my PR does.  Though the version distance isn't used by deoplete, it allows the tests to have results that can be confirmed with arithmetic on a napkin instead of having blind faith in boolean returns.

In any case, there are problems with your vimscript diff above:

- `deoplete#init#cmp_versions()` always returns `0`
  - The second argument of `split()` requires a pattern.  You're using `.` instead of `\.`.  So, you're comparing empty lists.
  - The function would still fail if the above item is fixed because you're comparing strings.
    - You are essentially doing what caused the bug in the first place.
  - This is the blind faith I spoke of.
- `deoplete#init#zip_longest()` does more than what is needed.
  - It's a new function to support another new function.  It isn't used elsewhere and it's not so complex that it improves readability.  It should just be in `deoplete#init#cmp_versions()`.
  - It's making copies of lists to fill in shorter lists.
    - One could assume that longer lists are automatically `>` shorter lists if the loop doesn't return early.
  - There shouldn't be a case where anything other than `0` is needed to fill the lists.  Having a default value argument is superfluous and furthers the case of this function not being needed.

I'm a little tired at the moment, but in terms of complexity, I think your implementation is *at least* `O(n^2)` while mine is `O(n)`.  Given what I was able to accomplish in a single loop, I would say that your implementation is overkill and less performant.

Here's the version parser I wrote a while ago that I used as basis of that PR: [helpful.vim#L205-L273](https://github.com/tweekmonster/helpful.vim/blob/77b91e51080c06c02d93e91dfe59d1f16e82a04a/autoload/helpful.vim#L205-L273).  Not only can the results be verified, it can be reversed back into a version string with a little math.  The performance impact is basically nil despite how it looks.  If I had copied it line for line into deoplete, I don't think anyone would notice its one time execution at startup.

**Comment by purpleP** [comment]
@tweekmonster 
My python code isn't supposed to work, it's just to show a concept. Nevertheless, I'd like to say that `zip_longest` is a very common function and it should be in any programmer's toolbox and be familiar to most people. So while everything you've said about it is true, I think it should give more readability here. In fact `zip_longest` isn't needed and just `zip` would suffice. But that's not the point. 

The concept is that all you need is to iterate both list (splitted strings) at the same time (I've copied ZipLongest from one of the vimscript libraries for that, but this can be done via using indices) and use only comparisons between the numbers.

The comparisons when compiled to asm should be 1 cpu cycle while pow is should be about 100 cpu cycles. That's why I've said that this is an overkill.

> ```
def cmp(v1, v2):
    return [int(x) for x in v1.split('.')] < [int(x) for x in v2.split('.')]
```

While it hasn't occurred to me that it's possible to compare lists like that in python, this allocates two unnecessary lists.

As to your arguments about tests and napkin and reversing back - it's not clear to me. If we want function that compares versions, why do we want to reverse anything? Shouldn't creating three tests cases (first version is higher, both versions is the same, second version is higher) be enough? So that arguments haven't convinced me, unless you would clarify this somehow.

So to summarize.

* Performance. I'll give a pseudocode for more performant variant. It creates only two lists and uses only simple comparisons.
```python
def (a, b):
    anums = a.split('.')
    bnums = b.split('.')
    alen = len(a)
    blen = len(b)
    len_cmp = cmp(alen, blen)
    min_len = alen if len_cmp < 0 else blen
    for i in range(min_len):
         x = cmp(int(alen[i]), int(alen[i])
         if x != 0:
             return x
    return len_cmp
```

The tests could be implemented as 4 tests cases.

- '0.1', '0.2'
- '0.2', '0.1'
- '0.1', '0.1'
- '0.1.1', '0.1.0'
